pkg/optimization/substat_details.go

```diff
@@ -0,0 +1,551 @@
+package optimization
+
+import (
+	"fmt"
+	"github.com/genshinsim/gcsim/internal/simulator"
+	"github.com/genshinsim/gcsim/pkg/core"
+	"math"
+	"sort"
+	"strings"
+)
+
+type SubstatOptimizerDetails struct {
+	charRelevantSubstats   map[core.CharKey][]core.StatType
+	artifactSets4Star      []string
+	substatValues          []float64
+	mainstatValues         []float64
+	charSubstatFinal       [][]int
+	charSubstatLimits      [][]int
+	charSubstatRarityMod   []float64
+	charProfilesInitial    []core.CharacterProfile
+	charWithFavonius       []bool
+	charProfilesERBaseline []core.CharacterProfile
+	charProfilesCopy       []core.CharacterProfile
+	simcfg                 core.SimulationConfig
+	simopt                 simulator.Options
+	cfg                    string
+	indivSubstatLiquidCap  int
+	totalLiquidSubstats    int
+}
+
+// Calculate per-character per-substat "gradients" at initial state using finite differences
+// In practical evaluations, adding small numbers of substats (<10) can be VERY noisy
+// Therefore, "gradient" evaluations are done in groups of 10 substats
+// Allocation strategy is to just max substats according to highest gradient to lowest
+// TODO: Probably want to refactor to potentially run gradient step at least twice:
+// once initially then another at 10 assigned liquid substats
+// Fine grained evaluations are too expensive time wise, but can perhaps add in an option for people who want to sit around for a while
+func (stats *SubstatOptimizerDetails) optimizeNonERSubstats() []string {
```

Comment:
if i understand correctly this is return the debug msgs. is there a reason why we're returning this instead of just printing it out?

Reply:
In addition, for both this function as well as optimizeER function. i feel like it'd be more logical if these functions returns optimal result given starting point (+ whatever other params). so a bit more "pure" functional if you will (not sure what the right wording is here)